src/provider/remoteFileSystemProvider.ts

```diff
@@ -359,29 +359,44 @@ class VirtualFileSystem {
         if (fileType && fileType==='doc' && fileEntity) {
             const doc = fileEntity as DocumentEntity;
             const _content = JSON.stringify(new TextDecoder().decode(content));
+            let currentPos = 0;
+            let hisCount = 0;
             if (doc.version && doc.cache) {
                 const update = {
                     doc: doc.name,
                     lastV: doc.lastVersion,
-                    v: doc.version + 1,
+                    v: doc.version,
                     // Reference: services/web/frontend/js/vendor/libs/sharejs.js#L1288
                     hash: (()=>{
                         if (!doc.mtime || Date.now()-doc.mtime>5000) {
                             doc.mtime = Date.now();
+                            const _content = new TextDecoder().decode(content);
                             return require('crypto').createHash('sha1').update(
-                                "blob " + doc.cache.length + "\x00" + doc.cache
+                                "blob " + _content.length + "\x00" + _content
                             ).digest('hex');
                         }
                     })() as string,
+                    
                     op: (()=>{
-                        return Diff.diffChars(doc.cache, _content)
+                        return Diff.diffChars(JSON.parse(doc.cache), JSON.parse(_content))
```

Comment:
If you are sure "raw string" is not needed, please help remove all the `parse/stringify` things in this file.